kong/async.lua

```diff
@@ -0,0 +1,164 @@
+local semaphore = require "ngx.semaphore"
+
+
+local ngx = ngx
+local kong = kong
+local pcall = pcall
+local select = select
+local unpack = unpack
+local setmetatable = setmetatable
+
+
+local QUEUE_SIZE = 100000
+
+
+local function get_pending(self)
+  local head = self.head
+  local tail = self.tail
+  if head < tail then
+    head = head + QUEUE_SIZE
+  end
+  return head - tail
+end
+
+
+local function job_thread(self, index)
+  while not ngx.worker.exiting() do
+    local ok, err = self.work:wait(1)
+    if ok then
+      if self.head ~= self.tail then
+        local tail = self.tail == QUEUE_SIZE and 1 or self.tail + 1
+        local job = self.jobs[tail]
+        self.tail = tail
+        self.jobs[tail] = nil
+        ok, err = job()
+        if not ok then
+          kong.log.err("async thread #", index, " job error: ", err)
+        end
+      end
+
+    elseif err ~= "timeout" then
```

Comment:
is there a special case to be defined if the worker process starts exiting while this thread is waiting for `self.work:wait(1)`, because currently it starts logging errors if I read this code correctly.

Reply:
I have never seen it starting to log errors in the case you said. I mean there are 100 threads on each worker and no error log into any logs with restarts or reloads. The `wait(1)` means `1 second at max`.